sycl/source/detail/error_handling/enqueue_kernel.cpp

```diff
@@ -116,120 +139,89 @@ bool oclHandleInvalidWorkGroupSize(const device_impl &DeviceImpl,
+                                       PI_INVALID_WORK_GROUP_SIZE);
+          else
+            throw sycl::nd_range_error(
+                "Global_work_size must be evenly divisible by local_work_size. "
+                "Non-uniform work-groups are not supported by the target "
+                "device",
+                PI_INVALID_WORK_GROUP_SIZE);
+        } else {
+          // OpenCL 2.x:
+          // PI_INVALID_WORK_GROUP_SIZE if the program was compiled with
+          // –cl-uniform-work-group-size and the number of work-items specified
+          // by global_work_size is not evenly divisible by size of work-group
+          // given by local_work_size
+
+          pi_program Program = nullptr;
+          Plugin.call<PiApiKind::piKernelGetInfo>(
+              Kernel, PI_KERNEL_INFO_PROGRAM, sizeof(pi_program), &Program,
+              nullptr);
+          size_t OptsSize = 0;
+          Plugin.call<PiApiKind::piProgramGetBuildInfo>(
+              Program, Device, PI_PROGRAM_BUILD_INFO_OPTIONS, 0, nullptr,
+              &OptsSize);
+          string_class Opts(OptsSize, '\0');
+          Plugin.call<PiApiKind::piProgramGetBuildInfo>(
+              Program, Device, PI_PROGRAM_BUILD_INFO_OPTIONS, OptsSize,
+              &Opts.front(), nullptr);
+          const bool HasStd20 =
+              Opts.find("-cl-std=CL2.0") != string_class::npos;
+          const bool RequiresUniformWGSize =
+              Opts.find("-cl-uniform-work-group-size") != string_class::npos;
+          std::string message =
+              LocalExceedsGlobal
+                  ? "Local workgroup size greater than global range size. "
+                  : "Global_work_size not evenly divisible by "
+                    "local_work_size. ";
+          if (!HasStd20)
+            throw sycl::nd_range_error(
+                message.append(
+                    "Non-uniform work-groups are not allowed by "
+                    "default. Underlying "
+                    "OpenCL 2.x implementation supports this feature "
+                    "and to enable "
+                    "it, build device program with -cl-std=CL2.0"),
+                PI_INVALID_WORK_GROUP_SIZE);
+          else if (RequiresUniformWGSize)
+            throw sycl::nd_range_error(
+                message.append(
+                    "Non-uniform work-groups are not allowed by when "
+                    "-cl-uniform-work-group-size flag is used. Underlying "
+                    "OpenCL 2.x implementation supports this feature, but it "
+                    "is "
+                    "being "
+                    "disabled by -cl-uniform-work-group-size build flag"),
+                PI_INVALID_WORK_GROUP_SIZE);
+          // else unknown.  fallback (below)
+        }
       }
     }
+    throw sycl::nd_range_error(
+        "Non-uniform work-groups are not supported by the target device",
+        PI_INVALID_WORK_GROUP_SIZE);
   }
```

Comment:
All of the checks above are done only for the OpenCL backend.  I presume something is needed for the other backends too?  If so, I'd recommend at least adding a TODO comment to that effect:

```
}
else {
    // TODO: Similar checks should be done for the non-OpenCL backends.
}
```

Here, the `else` should match the `if (Platform.get_backend() == cl::sycl::backend::opencl)` on line 143.